File: Plot/plot.py
import pandas as pd
import matplotlib.pyplot as plt
import os

def load_results(csv_path, label):
    df = pd.read_csv(
        csv_path,
        header=None,
        names=[
            "filename", "nodes", "delta", "constant", "colours_used",
            "total_time", "total_memory", "avg_list_size", "conflict_edges"
        ]
    )
    df["memory_MB"] = df["total_memory"] / (1024 * 1024)
    df["Delta+1"] = df["delta"] + 1
    df["label"] = label
    df["shortfile"] = df["filename"].apply(lambda x: os.path.basename(str(x)))
    return df

def plot_vs_deltap1(apst_df, greedy_df, metric, ylabel, filename):
    plt.figure(figsize=(8, 6))

    # Use only constant == 40 for APST
    apst_40 = apst_df[apst_df["constant"] == 40].sort_values("Delta+1")
    greedy_sorted = greedy_df.sort_values("Delta+1")

    plt.plot(apst_40["Delta+1"], apst_40[metric], marker='o', linestyle='-', label='APST (constant=40)')
    plt.plot(greedy_sorted["Delta+1"], greedy_sorted[metric], marker='s', linestyle='--', label='Greedy')

    # Annotate each point with the filename
    for i, row in apst_40.iterrows():
        plt.annotate(row["shortfile"], (row["Delta+1"], row[metric]), textcoords="offset points", xytext=(0,5), ha='center', fontsize=8)
    for i, row in greedy_sorted.iterrows():
        plt.annotate(row["shortfile"], (row["Delta+1"], row[metric]), textcoords="offset points", xytext=(0,-12), ha='center', fontsize=8, color='red')

    plt.xlabel("Delta + 1")
    plt.ylabel(ylabel)
    plt.title(f"{ylabel} vs Delta+1 (APST constant=40 vs Greedy)")
    plt.legend()
    plt.grid(True)
    plt.tight_layout()
    plt.savefig(filename)
    plt.close()
    print(f"Saved: {filename}")

if __name__ == "__main__":
    apst_file = "apst_results.csv"
    greedy_file = "greedy_results.csv"
    apst_df = load_results(apst_file, "APST")
    greedy_df = load_results(greedy_file, "Greedy")

    plot_vs_deltap1(apst_df, greedy_df, "colours_used", "Colours Used", "colours_vs_deltap1.png")
    plot_vs_deltap1(apst_df, greedy_df, "total_time", "Time (s)", "time_vs_deltap1.png")
    plot_vs_deltap1(apst_df, greedy_df, "memory_MB", "Memory (MB)", "memory_vs_deltap1.png")

File: Plot/plot.cpp
#include "../APST/apst.h"
#include "../Greedy-Algorithm/greedy_only.h"
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <iomanip>
#include <filesystem> 

int main() {
    std::string file = "../Graphs/large_twitch_edges.csv";
    std::string out_txt_path = "../Plot/results.txt"; 

    // Greedy algorithm 
    GreedyResult greedy_result = greedy(file);
    std::cout << "Greedy: colours used = " << greedy_result.colours_used 
              << ", success = " << (greedy_result.success ? "Yes" : "No") << std::endl;

    // APST algorithm for all constants
    std::vector<APSTResult> apst_results;
    for (int c = 30; c <= 40; ++c) {
        double constant = static_cast<double>(c);
        APSTResult result = apst(file, constant);
        apst_results.push_back(result);
    }

    // Write results table
    std::ofstream out(out_txt_path, std::ios::app);
    out << "filename constant nodes delta success apst_colours apst_time apst_memory greedy_colours greedy_time greedy_memory\n";
    std::string shortfile = file.substr(file.find_last_of("/\\") + 1);
    out << std::fixed << std::setprecision(4);

    for (const auto& result : apst_results) {
        out << shortfile << ' '
            << result.constant << ' '
            << result.nodes << ' '
            << result.delta << ' '
            << result.success << ' '
            << result.colours_used << ' '
            << result.total_time << ' '
            << result.total_memory << ' '
            << greedy_result.colours_used << ' '
            << greedy_result.total_time << ' '
            << greedy_result.total_memory << '\n';
    }

    out.close();
    return 0;
}
